apps/emqx/src/emqx_broker.erl

```diff
@@ -204,15 +199,12 @@ do_unsubscribe2(Topic, SubPid, SubOpts) when
     is_binary(Topic), is_pid(SubPid), is_map(SubOpts)
 ->
     _ = emqx_broker_helper:reclaim_seq(Topic),
-    case maps:get(shard, SubOpts, 0) of
-        0 ->
-            true = ets:delete_object(?SUBSCRIBER, {Topic, SubPid}),
-            emqx_exclusive_subscription:unsubscribe(Topic, SubOpts),
-            cast(pick(Topic), {unsubscribed, Topic});
-        I ->
-            true = ets:delete_object(?SUBSCRIBER, {{shard, Topic, I}, SubPid}),
-            cast(pick({Topic, I}), {unsubscribed, Topic, I})
-    end;
+    I = maps:get(shard, SubOpts, 0),
+    case I of
+        0 -> emqx_exclusive_subscription:unsubscribe(Topic, SubOpts);
+        _ -> ok
+    end,
+    cast(pick(Topic), {unsubscribed, Topic, SubPid, I});
```

Comment:
Just thinking aloud: a side effect of `cast(pick(Topic)` instead of `cast(pick({Topic, I})` is less even workload distribution over broker pool workers. 
Say, topic A has a very large number of subscribers. If all/most of A subscribers go down/unsubscribe, they will be processed by the same broker process, probably making it less responsive for a while.
Consistency-wise it's good and simple, though.. :thinking:

Reply:
Very good point. To be honest I'm struggling to come up with a better alternative that does not involve significant rework of route syncing logic. Maybe I'm missing something? From the correctness/consistency standpoint, my assumptions are:
1. Order of changes of each route is important, and should be preserved in the whole system: we achieve this by synchronizing and sending changes of the same topic from the same process.<sup>Though admittedly, mnesia might compromise it somewhere down the line.</sup>
1. If we are relying on local state to minimize global routing state changes, it's important that some operations on ETS tables on the same topic should not be interleaved (in this case, `ets:member` + `ets:insert` vs `ets:delete`s).

Reply:
I'm also thinking about alternatives and have not found any better solution either yet..

Reply:
alternative is to one route for each shard.

Reply:
> alternative is to one route for each shard.
not claiming it's a better solution though.